### ISSUE TITLE
Emtpy interface reduced to `never` in nested generic function 

### ISSUE THREAD

**Issue by panmenghan** [microsoft/TypeScript#49536]
# Bug Report
I can't find the docs about this behavior or google. I am not sure this is a bug.

My intention was define the `IEmpty` interface and augment it later, but the compiler reduce the `IEmpty` interface to `never`, my current workaround that have to define two dummy properties or not use nested function(or closure).

Is there a right way to force the compiler to keep the `IEmpty` in nested generic function?

### 🔎 Search Terms

closure, nested, generic, emtpy interface, never, narrow

### 🕗 Version & Regression Information

- This is the behavior from v4.2.3 to v4.7.3

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.2.3#code/PTAEFMFsAcBcE9QEsB2twCcBmBDAxuALABQq62+4oAkgKIwKgDeAviSeAB7QD2GsoLAFcUeWEh4pBIvAB4AKhE7oUAEwDOoANbh4PLDXpx4APgAUASmZtiHbnwHDR4yaEg4dAMRmXmJUNLOElIo4Oroqt6iCkoqGtq6+oYMpr6s-qAY4LBCGCFhEVF4JDbsxCDIaJi4BKAA7kiwABagklTQGDzQmAgkZNWUNADyoQAKnd38iEwZOABcoOEYqADmJWVcvPyBYsE7I+DjXT3wMVxxmjp6BtQHR5MI5lbptsSbDjsuUu5eMncTJzSGScu1coXC4EiMjOynAakuiRu-2OUye1gyWRyeVA4MKMnWrxIFX6FFqDWaoFgdR4oA6KN6pCqpKo1Hk1PuJz8xAC80WsGWKDW3NAACMFktVgS7FtHDIvjs2TwOVMYRcEtcaIrlY80qU3vZtiD5T9wEUtQCpkDhUa9rjIUVVXD4lckqz2Radc8bAFMbl8hCoaICUA)

### 💻 Code

```ts
// empty interface
interface IEmpty {}

export function func<T extends keyof IEmpty>() {}

export function makeFunc() {
  function nestedFunc<T extends keyof IEmpty>() {}
  return nestedFunc
}


// interface with one property
interface IOneProperty {
  a: string
}

export function funcOneProperty<T extends keyof IOneProperty>() {}

export function makeFuncOneProperty() {
  function nestedFunc<T extends keyof IOneProperty>() {}
  return nestedFunc
}


// interface with two properties
interface ITwoProperty {
  a: string
  b: string
}

export function funcTwoProperty<T extends keyof ITwoProperty>() {}

export function makeFuncTwoProperty() {
  function nestedFunc<T extends keyof ITwoProperty>() {}
  return nestedFunc
}
```

### 🙁 Actual behavior
`.D.TS` output:

```ts
interface IEmpty {
}
export declare function func<T extends keyof IEmpty>(): void;
export declare function makeFunc(): <T extends never>() => void;  // <-- unexpected: `keyof IEmpty` replaced by `never`
interface IOneProperty {
    a: string;
}
export declare function funcOneProperty<T extends keyof IOneProperty>(): void;
export declare function makeFuncOneProperty(): <T extends "a">() => void; // <-- same here
interface ITwoProperty {
    a: string;
    b: string;
}
export declare function funcTwoProperty<T extends keyof ITwoProperty>(): void;
export declare function makeFuncTwoProperty(): <T extends keyof ITwoProperty>() => void;
export {};
```
### 🙂 Expected behavior
`.D.TS` output:

```ts
interface IEmpty {
}
export declare function func<T extends keyof IEmpty>(): void;
export declare function makeFunc(): <T extends IEmpty>() => void;
interface IOneProperty {
    a: string;
}
export declare function funcOneProperty<T extends keyof IOneProperty>(): void;
export declare function makeFuncOneProperty(): <T extends IOneProperty>() => void;
interface ITwoProperty {
    a: string;
    b: string;
}
export declare function funcTwoProperty<T extends keyof ITwoProperty>(): void;
export declare function makeFuncTwoProperty(): <T extends keyof ITwoProperty>() => void;
export {};
```

**Comment by MartinJohns** [comment]
Sounds like a duplicate of #17294 / #27171.

**Comment by siarheiyelin** [comment]
Hello,

Unfortunately, I was unable to make the workaround above work (suggested by @weswigham ```keyof (IEmpty & {})```).

I have the same issue in a slightly different use case. There is a TS project which contains the code like this:
```typescript
type DistributiveOmit<T, K extends keyof any> = T extends any ? Omit<T, K> : never;

interface IProps {
    size: '1' | '2';
}
// It is empty during build time, but will be changed via module augmentation by external code
export interface IEmpty {}

type MyType = DistributiveOmit<IProps, keyof IEmpty> & IEmpty;
function withMods<T>() {
    // just an example
    return null as (T & { a: 1 });
}

export const Button = withMods<MyType>();
```
My goal is to compile the project, generate DTS, and then use module augmentation in a separate TypeScript project to dynamically modify the `IEmpty` interface.

**Actual behavior**
After compilation, the corresponding DTS includes a type definition for the Button, where the `keyof IEmpty` statement has been replaced with `never`. This makes it impossible to use module augmentation to modify the `IEmpty` interface. And therefore, the type of `Button` which depends on `IEmpty` will always remain unchanged. 

Code sandbox: [link](https://www.typescriptlang.org/play/?strict=false&noImplicitAny=false&strictNullChecks=false&module=1&ts=5.4.5&ssl=3&ssc=1&pln=3&pc=10#code/C4TwDgpgBAIglgZ2AJzgIwK7DgNwgeQFs5gAeAFQBooBpKCAD2AgDsATBKAawhAHsAZlACGLEAD4oAXijl6TVhxFioAfihESFajUkAuKCwh5kAbgBQ5uC2bIBwgMbQAkgAVkfMJwDe5qP6gEOAAvCAMAcgBGcKgAHyhwgCZwiwBfcwB6DKhnYChEekIwUCg2DFQWAHMoTDgAGzYobEIIakw8gHd6uproBwALUUqIRpw4YShCPjK66GEMSpabYWw+FhqQeVsWYR6HaYhzRjA+ZDzrW3snHIBRIpLvdPNQSCgAWRBycGgZeCRUdq4AjEMhuDxeag8fhCZx3YoSKAAMlu9xAFgEGBYDlW6y6wH6b2mCAo4gAFABKKC+AJQLJQABWGCQynkwiKsz8AWQEGA5XWLAwdR6wk4pLkyO8IgMkSgqXJaUsx1OeX2LGZACEsMA1tIoHiCUTSB8vpAyfKgA)

**Expected behavior**
Is there a way to explicitly prevent this behavior and preserve `keyof IEmpty` statement in the above use cases?

Thanks in advance.

**Comment by panmenghan** [comment]
Fixed in v5.5.2 😄
